api/server/runner_httptrigger.go

```diff
@@ -112,15 +112,23 @@ func (trw *triggerResponseWriter) WriteHeader(statusCode int) {
 		realHeaders[k] = vs
 	}
 
-	// XXX(reed): simplify / add tests for these behaviors...
-	gatewayStatus := 200
-	if statusCode >= 400 {
-		gatewayStatus = 502
-	} else if fnStatus > 0 {
-		gatewayStatus = fnStatus
+	trw.inner.WriteHeader(determineFinalStatus(statusCode, fnStatus))
+}
+
+func determineFinalStatus(serviceStatus, userFnStatus int) int {
+	finalStatus := 200
+
+	if userFnStatus > 0 {
+		if userFnStatus >= 500 {
+			finalStatus = 502
+		} else {
+			finalStatus = userFnStatus
+		}
+	} else if serviceStatus > 0 {
+		finalStatus = serviceStatus
```

Comment:
thanks for extracting func on this. big readability improvement for me would be to return statuses instead of falling through a bunch of branches (always feels like a newspaper maze puzzle to me anyway), also I just wanted to try the new suggestion thing, here goes: 

```suggestion
func determineFinalStatus(serviceStatus, userFnStatus int) int {
                switch {
                case userFnStatus > 400:
                  return 502
                case userFnStatus > 0:
                  return userFnStatus
                case serviceStatus > 0:
                  return serviceStatus
                default:
                  return 200
                }
}
```

Reply:
hmm, need some way to suggest over a range of lines, cool idea though something I didn't know I wanted (I suggest a lot)

Reply:
Ooo i like that. Yeah this was well fugly, was more interested in the what over how. I way prefer your how.